### PR TITLE
Channel posts and edited channel posts support

### PR DESCRIPTION
**Pull request by MarcoBuster**
With new [Telegram Bot API Update (2.3)](https://core.telegram.org/bots/api#recent-changes) now bots can receive messages _(updates)_ from **channels**
There are two types of updates:

- **Channel post**: returns `Chat` and `Message` objects when the channel posts a new message.
You can process this by doing:
```
@bot.channel_post
def channel_post(chat, message):
    # things
```
- **Edited channel post**: returns `Chat` and `Message` objects when a channel post was edited
You can process this by doing:
```
@bot.edited_channel_post
def edited_channel_post(chat, message):
    # things
```

### CODE REVIEW

**Comment by pietroalbini** [comment]
You're failing the lints, which checks if the coding style (pep8) is respected. You can run `invoke lint` to check for those errors yourself.

More information about the coding style: [pep8.org](http://pep8.org).

File: botogram/bot.py [hunk 1]
```diff
@@ -144,6 +148,16 @@ def message_edited(self, func):
         self._main_component.add_message_edited_hook(func)
         return func
 
+    def channel_post(self, func):
+        """Add a channel post hook"""
+        self._main_component.add_channel_post_hook(func)
+        return func
+
+    def edited_channel_post(self, func):
```
Comment (pietroalbini) [inline botogram/bot.py]: Maybe it's better to call this `channel_post_edited`, for the sake of consistency with `message_edited`.

File: botogram/components.py [hunk 2]
```diff
@@ -158,6 +160,22 @@ def add_message_edited_hook(self, func):
...
 
+    def add_channel_post_hook(self, func):
+        """Add a channel post hook"""
+        if not callable(func):
+            raise ValueError("A channel post hook must be callable")
+
+        hook = hooks.ChannelPostHook(func, self)
+        self.__channel_post_hooks.append(hook)
+
+    def add_edited_channel_post_hook(self, func):
```
Comment (pietroalbini) [inline botogram/components.py]: Maybe it's better to call this `add_channel_post_edited_hook`, for the sake of consistency with `add_message_edited_hook`.

File: botogram/components.py [hunk 3]
```diff
@@ -158,6 +160,22 @@ def add_message_edited_hook(self, func):
...
+        """Add a channel post hook"""
```
Comment (pietroalbini) [inline botogram/components.py]: You probably missed a docstring here.

File: botogram/hooks.py [hunk 4]
```diff
@@ -211,6 +211,21 @@ def _call(self, bot, update):
...
 
+class ChannelPostHook(Hook):
+    """Underlying hook for @bot.channel_post"""
+    
+    def _call(self, bot, update):
+        message = update.channel_post
+        return bot._call(self.func, self.component_id, chat=message.chat,
+                         message=message)
+
+class EditedChannelPostHook(Hook):
```
Comment (pietroalbini) [inline botogram/hooks.py]: Maybe it's better to call this `ChannelPostEditedHook`, for the sake of consistency with `MessageEditedHook`.

File: botogram/messages.py [hunk 5]
```diff
@@ -37,3 +37,33 @@ def process_edited_message(bot, chains, update):
...
+                             (update.update_id, hook.name))
+            return
+
+    bot.logger.debug("No hook actually processed the #%s update." %
+                     update.update_id)
+    
+def process_edited_channel_post(bot, chains, update):
+    """Process an edited channel post"""
+    for hook in chains["edited_channel_post"]:
+        bot.logger.debug("Processing edited_channel_post in update #%s with the "
```
Comment (pietroalbini) [inline botogram/messages.py]: Why those underscores?

File: botogram/messages.py [hunk 6]
```diff
@@ -37,3 +37,33 @@ def process_edited_message(bot, chains, update):
 
     bot.logger.debug("No hook actually processed the #%s update." %
                      update.update_id)
+
+def process_channel_post(bot, chains, update):
+    """Process a channel post"""
+    for hook in chains["channel_post"]:
+        bot.logger.debug("Processing channel_post in update #%s with the "
```
Comment (pietroalbini) [inline botogram/messages.py]: Why this underscore?

**Comment by MarcoBuster** [comment]
Now all is fine!

File: botogram/components.py (hunk 3 above)
Comment (MarcoBuster) [inline botogram/components.py]: Ops

File: botogram/messages.py (hunk 6 above)
Comment (MarcoBuster) [inline botogram/messages.py]: Ops

File: botogram/bot.py (hunk 1 above)
Comment (MarcoBuster) [inline botogram/bot.py]: Ok!

File: botogram/hooks.py (hunk 4 above)
Comment (pietroalbini) [inline botogram/hooks.py]: This is not outdated btw

File: docs/api/bot.rst [hunk 7]
```diff
@@ -194,6 +194,50 @@ components.
...
+
+      .. code-block:: python
+
+        @bot.channel_post
+        def channel_post(chat, message):
+            message.reply("This post is cool")
+
+      .. versionadded:: 0.3
+
+    .. py:decoratormethod:: channel_post
```
Comment (paolobarbolini) [inline docs/api/bot.rst]: This is the same as the previous one

File: docs/api/bot.rst [hunk 8]
```diff
@@ -194,6 +194,50 @@ components.
...
+      * **message**: the message (instance of
+        :py:class:`~botogram.Message`)
+
+      .. code-block:: python
+
+        @bot.channel_post
+        def channel_post(chat, message):
+            message.reply("This post is cool")
+
+      .. versionadded:: 0.3
```
Comment (paolobarbolini) [inline docs/api/bot.rst]: The correct versionadded is 0.4

File: docs/api/bot.rst [hunk 9]
```diff
@@ -194,6 +194,50 @@ components.
...
+      * **message**: the (new) edited message (instance of
+        :py:class:`~botogram.Message`)
+
+      .. code-block:: python
+
+        @bot.edited_channel_post
+        def edit_channel_post(chat, message):
+            message.reply("Why you edited this post? It was so cool :(")
+
+      .. versionadded:: 0.3
```
Comment (paolobarbolini) [inline docs/api/bot.rst]: The correct versionadded is 0.4

File: docs/api/bot.rst [hunk 10]
```diff
@@ -194,6 +194,50 @@ components.
...
+
+        @bot.channel_post
+        def channel_post(chat, message):
+            message.reply("This post is cool")
+
+      .. versionadded:: 0.3
+
+    .. py:decoratormethod:: channel_post
+
+      This decorator receive messages from channels, with new Telegram Bot API update
```
Comment (paolobarbolini) [inline docs/api/bot.rst]: You haven't specified that this one receives edits

**Comment by pietroalbini** [comment]
Merged!

**Pull request closed without merging**